nexus/tests/integration_tests/internet_gateway.rs

```diff
@@ -0,0 +1,535 @@
+        create_instance_with, create_internet_gateway, create_ip_pool,
+        create_project, create_route, create_router, create_vpc,
+        delete_internet_gateway, detach_ip_address_from_igw,
+        detach_ip_pool_from_igw, link_ip_pool, objects_list_page_authz,
+    },
+};
+use nexus_test_utils_macros::nexus_test;
+use nexus_types::external_api::{
+    params::{
+        ExternalIpCreate, InstanceNetworkInterfaceAttachment,
+        InstanceNetworkInterfaceCreate,
+    },
+    views::{InternetGateway, InternetGatewayIpAddress, InternetGatewayIpPool},
+};
+use nexus_types::identity::Resource;
+use omicron_common::{
+    address::{IpRange, Ipv4Range},
+    api::external::{
+        IdentityMetadataCreateParams, NameOrId, RouteDestination, RouteTarget,
+    },
+};
+
+type ControlPlaneTestContext =
+    nexus_test_utils::ControlPlaneTestContext<omicron_nexus::Server>;
+
+const PROJECT_NAME: &str = "delta-quadrant";
+const VPC_NAME: &str = "dominion";
+const IGW_NAME: &str = "wormhole";
+const IP_POOL_NAME: &str = "ds9";
+const IP_POOL_ATTACHMENT_NAME: &str = "runabout";
+const IP_ADDRESS_ATTACHMENT_NAME: &str = "defiant";
+const IP_ADDRESS_ATTACHMENT: &str = "198.51.100.47";
+const IP_ADDRESS_ATTACHMENT_FROM_POOL: &str = "203.0.113.1";
+const INSTANCE_NAME: &str = "odo";
+const FLOATING_IP_NAME: &str = "floater";
+const ROUTER_NAME: &str = "deepspace";
+const ROUTE_NAME: &str = "subspace";
+
+#[nexus_test]
+async fn test_internet_gateway_basic_crud(ctx: &ControlPlaneTestContext) {
+    let c = &ctx.external_client;
+    test_setup(c).await;
+
+    // should start with just default gateway
+    let igws = list_internet_gateways(c, PROJECT_NAME, VPC_NAME).await;
+    assert_eq!(igws.len(), 1, "should start with zero internet gateways");
+    expect_igw_not_found(c, PROJECT_NAME, VPC_NAME, IGW_NAME).await;
+
+    // create an internet gateway
+    let gw = create_internet_gateway(c, PROJECT_NAME, VPC_NAME, IGW_NAME).await;
+    let igws = list_internet_gateways(c, PROJECT_NAME, VPC_NAME).await;
+    assert_eq!(igws.len(), 2, "should now have two internet gateways");
+
+    // should be able to get the gateway just created
+    let same_igw = get_igw(c, PROJECT_NAME, VPC_NAME, IGW_NAME).await;
+    assert_eq!(&gw.identity, &same_igw.identity);
+
+    // a new igw should have zero ip pools
+    let igw_pools =
+        list_internet_gateway_ip_pools(c, PROJECT_NAME, VPC_NAME, IGW_NAME)
+            .await;
+    assert_eq!(igw_pools.len(), 0, "a new igw should have no pools");
+
+    // a new igw should have zero ip addresses
+    let igw_addrs =
+        list_internet_gateway_ip_addresses(c, PROJECT_NAME, VPC_NAME, IGW_NAME)
+            .await;
+    assert_eq!(igw_addrs.len(), 0, "a new igw should have no addresses");
+
+    // attach an ip pool
+    attach_ip_pool_to_igw(
+        c,
+        PROJECT_NAME,
+        VPC_NAME,
+        IGW_NAME,
+        IP_POOL_NAME,
+        IP_POOL_ATTACHMENT_NAME,
+    )
+    .await;
+    let igw_pools =
```

Comment:
If we're intending to allow multiple pools/addrs to be linked to an IGW, I think we should also be testing that in here.

Reply:
I was not intending for this to be the case, but it actually seems like a good constraint on this initial implementation/PR. When we have a need to move to more than one IP pool association per IGW we can add a lot more testing around that use case.

Reply:
Okay, that sounds reasonable. Thanks for clarifying!